src/components/hooks/useStructure.js

```diff
@@ -0,0 +1,31 @@
+import React, { useState, useEffect, createContext } from "react";
+
+import * as ContentProvider from "../services/contentProvider";
+
+export const structureContext = createContext();
+
+export function StructureProvider({ children }) {
+  const [structure, setStructure] = useState([]);
+
+  useEffect(() => {
+    const unsubscribe = ContentProvider.getStructure({
+      next: (querySnapshot) => {
+        let appStructure = [];
+        let structure = querySnapshot.docs;
+        Promise.all(
+          structure.map((category) => appStructure.push(category.data()))
+        ).then(() => {
+          setStructure(appStructure);
+        });
```

Comment:
While this construct works nicely, I believe it's quite hard to read. Personally I would prefer keeping just the promise array in the `Promise.All` function call,
Like so: `structure.map((doc) => doc.data()`
Then applying the results:
```JavaScript
Promise.all(structure.map((doc) => doc.data())
.then((results) => {
   setStructure(results);
}
```
This would also make the appStructure array unnecessary.

Disclaimer: Not sure if this code works. Wrote it on an iPad during a train ride 😄